Initial work on a refract parser and new SDK API

danielgtaylor opened the pull request:
This is an attempt to create a [refract](https://github.com/refractproject/refract-spec)-based new SDK interface. It looks something like this at the moment:

``` js
import fury from 'fury';

let api = fury.load(elements);

// Example output
// --------------
// Group Questions
// Resources related to questions in the API.
// Resource Question /questions/{id}
// Transition View a Questions Detail
// Transaction undefined
// Response 200 { ... some JSON body ... }

for (let group of api.resourceGroups) {
  console.log(`Group ${group.title}`);
  console.log(group.description);

  for (let resource of group.resources) {
    console.log(`Resource ${resource.title} ${resource.href}`);

    for (let transition of resource.transitions) {
      console.log(`Transition ${transition.title}`);

      for (let transaction of transition.transactions) {
        let response = transaction.response;
        console.log(`Transaction ${transaction.title}`);
        console.log(`Response ${response.statusCode} ${response.messageBody}`);
      }
    }
  }
}
```

Also included are some updates to use ES6 alongside the existing ES5 and Coffeescript, as well as adding a linter and fixing a few lint issues in the existing code (like trailing spaces or missing semicolons).

danielgtaylor committed:
Initial work on a refract parser and new SDK API

danielgtaylor commented on src/refract/element.es6:
```diff
@@ -0,0 +1,92 @@
+   *
+   * Returns an array of element objects.
+   */
+  contents(options={}) {
```
There's probably a **lot** more work to be done here as it currently only supports simple arrays of elements we care about or simple hashes where a key contains such an array.

kylef commented on src/refract/element.es6:
```diff
@@ -0,0 +1,92 @@
+ */
+const rawElement = Symbol('element');
+
+export default class RefractElement {
```
Perhaps it would make sense to use @smizell's [minim](https://github.com/smizell/minim) library.

danielgtaylor replied:
@kylef at the moment it doesn't seem to support metadata, custom element types or reading from compact refract. The first two are necessary for my use case. @smizell, should I be updating minim or do you have any plans around those features at the moment?

smizell replied:
@danielgtaylor I have plans to support everything in the Refract spec, but two things have been holding me back.
1. The spec has changed since starting minim, and it needs some work to get it up to speed (hence no support for metadata and so on)
2. I only get half a day per week to work on this, so I will be slow

With that said, if you want to update it and open PRs to it, that's totally fine. I had hoped to be able to convert to ES6 or vanilla JS.

danielgtaylor replied:
@smizell no problem, I'm working on updating it and building my code on top. I'm hoping to get something working and submitted tomorrow.

danielgtaylor committed:
Refract updates and tests

danielgtaylor commented on src/refract/primitives.es6:
```diff
@@ -0,0 +1,322 @@
+/*
+ * A refract element implementation with an extensible type registry.
```
@smizell I've vendored this for the time being because it feels wrong to open a pull request without updating the tests :tired_face:. I want this to be temporary until minim.js is updated, which will still require a good bit of work. My changes:
- [x] Convert to ES6
- [x] Remove `PropertyType`
- [x] Add `meta` to elements
- [x] Add a `fromCompactRefract` to support both input formats
- [x] Renamed a few variables to help non-native English speakers / readability (e.g. `val` -> `value`)
- [x] Added a global type registry (see `src/refract/api.es6` in this PR for how it is used)

Remaining work:
- [x] Support refracted elements within `attributes`, e.g. `transaction.request.headers`
- [ ] Better error handling

Open questions:
1. Should `ArrayType().push` take in refracted type instances or plain js values?
2. Should `ObjectType().set` take in refracted type instances or plain js values?
3. Should the type registry really be global?

smizell replied:
> Should ArrayType().push take in refracted type instances or plain js values?
> Should ObjectType().set take in refracted type instances or plain js values?

We could probably allow for both in some way, I'd say. Can we do this? In other words, if the type received is an instance, just store it, otherwise, convert it. Or have some options argument for `push`, `set`, etc. that allows you say not convert the value argument.

danielgtaylor replied:
I think I'd prefer to default to not touching the value (i.e. assume it is already refracted). It's how the array type works today, and how I think the object type and other nested types should work. The contract is simple to understand and consistent: if you need to push or set a simple type then call `convertToType` first. If you need to set a more complex element, then use your element class and construct it yourself. I'd like to try that out and if it becomes a pain point then revisit this discussion.

Btw, I've pushed up support for attributes which are themselves elements (e.g. Resource.hrefVariables in the resource namespace). It allows custom types to register key names in the attributes hash that should be converted to types. Let me know what you think of it. There are some examples and a simple test in the latest commit.

smizell replied:
I wanted to circle back and add some comments along the way.

> Should the type registry really be global?

I wanted to put a couple things down for you to think about here. We had the concept of this registry where you registered your own types, but is there a way to do this so you don't have to create new type classes? In other words, if an instance of a Refract element knows it's a string, it will always behave like a string no matter name you give it. 

The idea of a registry gets even more complicated when you consider namespaces. Really, the element definitions propogate down through the tree, but not up. This means that if you include a namespace, the parent elements have no visibility of those definitions, though all the children will.

> if you need to push or set a simple type then call convertToType first. If you need to set a more complex element, then use your element class and construct it yourself.

Minim started out as a simple interface to JSON structures, and Refract was born from that. It makes for going from JSON to Refract easy.

``` javascript
var obj = new Object
  .set('foo', 'bar')
  .set('a', 'b')
```

That makes sense to people new to these concepts. Having to include some convert function everywhere makes it somewhat less appealing to people coming from JSON, which uses simple dot notation.

I don't have any answers at this point, but I feel like getting adoption with this will require an interface that adds just a little complexity to what people are familiar with in the JSON world while providing the extra stuff Refract provides.

> Better error handling

This is one that I think has some really cool possibilities. My initial thought was to have an `Error` class that had all of the methods from all of the different kinds of elements (since JavaScript currently has no method missing). If you received an error, you created an instance and the rest of the chain worked, it just passed the error object all the way through. See this [PR here](https://github.com/smizell/minim/pull/21).

Kyle has brought up some interesting ideas around this, which would be to use Promises or [streams](https://github.com/jlongster/js-csp). This would be a change, but could provide an interface that can handle errors quite well.

---

Lots of rambling here. Hope it helps some! Excited to see the work and thought being put into this.

danielgtaylor replied:
@smizell thanks for the feedback!

> We had the concept of this registry where you registered your own types, but is there a way to do this so you don't have to create new type classes? In other words, if an instance of a Refract element knows it's a string, it will always behave like a string no matter name you give it.

You don't strictly have to create your own classes. The way it works now you can do this:

``` js
import {StringType, TypeRegistry} from 'primitives';

// I have a few elements I want to register that act like strings
TypeRegistry.elementMap['my-element'] = StringType;
TypeRegistry.elementMap['my-other-element'] = StringType;
TypeRegistry.elementMap.frob = StringType;

// Now I can handle them!
import {fromCompactRefract} from 'primitives';

let elements = fromCompactRefract(['array', {}, {}, [
  ['my-element', {}, {}, 'my-value'],
  ['my-other-element', {}, {}, 'my-other-value'],
  ['frob', {}, {}, 'abc123']
]);

console.log(elements.get(0).content); // => 'my-value'
```

> I don't have any answers at this point, but I feel like getting adoption with this will require an interface that adds just a little complexity to what people are familiar with in the JSON world while providing the extra stuff Refract provides.

I agree. Let me see what I can come up with on this front. It seems like my initial thinking was wrong and we do want to support built-in types easily.

The error PR is pretty interesting as well. I'm not sure how best to handle errors yet, but that sounds like a promising approach.

danielgtaylor commented:
@zdne @kylef please have another look. Ive updated the original description to reflect the latest code. This provides a new SDK API to interface with a refract-element based API description, and my plan would be to add new adapters that transform various formats into refract elements. For API Blueprint, this is hopefully just a single call out to Drafter. I'd love to get some initial feedback before I go too much further.

danielgtaylor committed:
Implement support for attributes which are elements

zdne commented:
@danielgtaylor could you please reflect the proposal in README ? I think it may be a good exercise to try describing the new – proposed interface in a human readable form.

zdne commented:
@danielgtaylor also I would like to see some explanation of the overall concept – is API (description) accessible only via ES6 classes or is there a direct access to elements?

zdne commented:
@danielgtaylor also, if possible I would say that the API shouldn't be assume certain nesting. E.g. you can have, in theory, multiple nested groups: 

```
- group A
    - group A.1
        - resource 1
    - group A.2
        - group A.2.i
            - resource 2
```

Possible?

zdne commented:
Related https://github.com/refractproject/refract-spec/pull/19

danielgtaylor commented:
@zdne I'll try and update the readme soon. I've been waiting to write a lot of the docs and tests until the refract namespaces are a bit more concrete :wink:.

You do not need ES6 or classes to use refract. Plain ol' `js` works fine:

``` javascript
var api = fury.load(elements);

// In the future:
var api = fury.parse('FORMAT: 1A\n\n# My API\n....');
```

Every element has its raw data available, as well as convenience accessors:

``` javascript
# Raw access to every element
console.log(api.element);    // => 'api'
console.log(api.meta);       // => {title: 'My API'}
console.log(api.attributes); // => {}
console.log(api.content);    // => [ResourceGroup(...), ResourceGroup(...), ...]

# Per-element conveniences
console.log(api.title);      // => 'My API'
api.title = 'New title';
console.log(api.meta.title); // => 'New title'
```

Since any `category` element can contain resource groups and a resource group is itself a subclass of a category that means your nested group example does indeed work:

``` javascript
let api = fury.load(elements);
console.log(api.resourceGroups[0].resourceGroups[1].resourceGroups[0].title); // => 'A.2.i'
```

I probably won't push an update to this until Monday. Please let me know if you have other questions/concerns!

danielgtaylor committed:
Update to latest refract spec.

Also adds new info to the README, fixes a couple minor issues and makes it
easier to filter key/value pairs such as those found in the HTTP headers.

danielgtaylor commented:
@zdne I've updated the readme in the latest push. Any feedback on the interface and documentation so far?

danielgtaylor committed:
Add a length method to StringType

danielgtaylor committed:
Add transition.attr getter

danielgtaylor committed:
Convert fury.js to ES6, add parse method stub

danielgtaylor committed:
Add an API Blueprint adapter

zdne commented on src/fury-emitter.js:
```diff
@@ -2,6 +2,7 @@ var events = require('events');
 
 // Default logging function
 function log(message) {
+  /*eslint no-unused-vars: 0 */
```
Not sure what is this for. Is this a package (dev)dependency?

danielgtaylor replied:
This just tells the linter to ignore unused variables in that function, i.e. do not fail the build. We want to keep the unused variable as an example to others writing their own loggers.

zdne commented:
@danielgtaylor looks good to me! 

Question: If I am right it is up to the adapter to decide whether it understands the format using the `detect()` correct? If so I would put some emphasis on this in the README. 

Additionally I am not 100% sure whether this will work for all the formats – What about relying on media type (e.g. `text/vnd.apiblueprint`). Adapter could register recognized media types and the `fury.parse()` would take a media type as additional parameter.

Another not, regarding README I would add a link to API namespace where appropriate (and maybe a bit of explanation) as this is basically a convenience interface to it. Contrary I would drop references to other namespaces (but I leave this one up to you).

danielgtaylor committed:
Update to latest minim

danielgtaylor commented:
Until the minim change is merged this will fail on Travis! refractproject/minim#23

smizell commented:
@danielgtaylor I've merged https://github.com/refractproject/minim/pull/23.

zdne commented:
@danielgtaylor what is the status on this?

danielgtaylor commented:
@zdne there seem to be some build/test issues I'm dealing with, then I'll push up the latest and it should be ready to merge.

zdne commented:
OK, thanks! 

> On May 18, 2015, at 11:04 AM, Daniel G. Taylor notifications@github.com wrote:
> 
> @zdne https://github.com/zdne there seem to be some build/test issues I'm dealing with, then I'll push up the latest and it should be ready to merge.
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/apiaryio/fury.js/pull/16#issuecomment-103150806.

danielgtaylor committed:
Move adapters into separate PR

danielgtaylor commented:
@zdne this should now be ready to be merged. Once in, I'll open a new PR for the adapter interface since I'm still making some changes to that.

pksunkara commented on README.md:
```diff
@@ -30,10 +30,73 @@ or as a dependency:
+}
+
+console.log('All API request URIs:');
+api.find(filterFunc).forEach(function (request) {
```
Should have an empty line before this to make this more readable.

pksunkara commented on src/refract/util.es6:
```diff
@@ -0,0 +1,45 @@
+ */
+function hasClass(element, name) {
+  return element.meta && element.meta.class &&
+         element.meta.class.indexOf &&
```
Maybe use `typeOf element.meta.class === 'string'` here?

danielgtaylor replied:
This is actually an array, not a string, but I can update that check.

``` js
{
  "element": "category",
  "meta": {
    "class": ["resourceGroup"]
  },
  "content": [...]
}
```

danielgtaylor replied:
Looks like the recommended way to check for an array is instead to use `Array.isArray(value)`, so I'll do that instead.

pksunkara commented:
Reviewed. Only 2 minor things.

danielgtaylor committed:
Upgrade eslint to fix no-shadow bug

danielgtaylor committed:
Code review feedback

danielgtaylor commented:
@pksunkara the items in your comments should be fixed now, thanks!
